packages/gatsby/src/internal-plugins/partytown/gatsby-ssr.tsx

```diff
@@ -0,0 +1,46 @@
+import React from "react"
+import type { GatsbySSR } from "gatsby"
+import { Partytown } from "@builder.io/partytown/react"
+import { PartytownContext } from "gatsby-script"
+import type { PartytownProps } from "@builder.io/partytown/react"
+
+const collectedScripts: Record<string, Array<PartytownProps>> = {}
+
+export const wrapRootElement: GatsbySSR[`wrapRootElement`] = ({
+  element,
+  pathname,
+}) => (
+  <PartytownContext.Provider
+    value={{
+      collectScript: (newScript: PartytownProps): void => {
+        collectedScripts[pathname] = [
+          ...(collectedScripts?.[pathname] || []),
+          newScript,
+        ]
+      },
+    }}
+  >
+    {element}
+  </PartytownContext.Provider>
+)
+
+export const onRenderBody: GatsbySSR[`onRenderBody`] = ({
+  pathname,
+  setHeadComponents,
+}) => {
+  if (!collectedScripts?.[pathname]?.length) {
+    return
+  }
+
+  const collectedForwards: Array<string> = collectedScripts?.[pathname]?.reduce(
+    (forwards: Array<string>, script: PartytownProps) => {
+      if (script?.forward) {
+        forwards = [...forwards, ...script?.forward]
```

Comment:
```suggestion
        forwards.push(script.forward)
```

to avoid initting arrays a lot

I would also think about using `Set` instead of array as accumulator which would get rid of duplicates (and then `forward={Array.from(scripts)}` if that prop expect array concretely and not just any iterable